### PR TITLE
[ReplicatedLoglet] Handle Append messages

### PR DESCRIPTION
**Pull request by muhamadazmy**
[ReplicatedLoglet] Handle Append messages

Summary:
Handle requests from remote loglets to append records

### CODE REVIEW

**Review by AhmedSoliman — CHANGES_REQUESTED** [review]

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 1]
```diff
@@ -11,27 +11,40 @@
...
+use crate::loglet::util::TailOffsetWatch;
+use crate::loglet::{AppendError, LogletCommit};
 
 type MessageStream<T> = Pin<Box<dyn Stream<Item = Incoming<T>> + Send + Sync + 'static>>;
 
-pub struct RequestPump {
+pub struct RequestPump<T> {
     metadata: Metadata,
     append_stream: MessageStream<Append>,
+    active_loglet: DashMap<ReplicatedLogletId, Arc<ReplicatedLoglet<T>>>,
```
[the remark this answers is not in the document]
Comment (muhamadazmy) [reply crates/bifrost/src/providers/replicated_loglet/network.rs:41]: The reason I did this, is that I need to get the Loglet Params from the chain. So I thought we could then cache it with the loglet id. 

Same reason I get the `last` segment of the logs on `get` otherwise we will have to scan the chain everytime to land on the correct segment index before i can fetch the params.

[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/replicated_loglet/network.rs:41]: Don't we send the index and log-id in the append message header?

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 2]
```diff
@@ -43,25 +56,183 @@ impl RequestPump {
...
+        append: &Incoming<Append>,
+    ) -> Result<(), SequencerStatus> {
+        let loglet = self
+            .get_latest_segment(provider, &append.body().header)
+            .await?;
+
+        // We get the sequencer instead of calling loglet.enqueue_batch to make sure this
+        // is the sequencer node. Otherwise, the client gets an error and should retry against
+        // the sequencer node.
+        let sequencer = loglet.sequencer().ok_or(SequencerStatus::NotSequencer)?;
```
[the remark this answers is not in the document]
Comment (muhamadazmy) [reply crates/bifrost/src/providers/replicated_loglet/network.rs:163]: No, you are right!

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 3]
```diff
@@ -43,25 +56,183 @@ impl RequestPump {
...
+                    known_global_tail: Some(global_tail.latest_offset()),
+                    sealed: Some(global_tail.is_sealed()), // this must be true
+                    status: SequencerStatus::Sealed,
+                },
+                first_offset: LogletOffset::INVALID,
+            },
+            Err(err) => Appended {
+                header: CommonResponseHeader {
+                    known_global_tail: Some(global_tail.latest_offset()),
+                    sealed: Some(global_tail.is_sealed()), // this must be true
```
[the remark this answers is not in the document]
Comment (muhamadazmy) [reply crates/bifrost/src/providers/replicated_loglet/network.rs:229]: No, copied by mistake

File: crates/bifrost/src/providers/replicated_loglet/loglet.rs [hunk 4]
```diff
@@ -121,6 +121,14 @@ impl<T: TransportConnect> ReplicatedLoglet<T> {
             log_server_manager,
         })
     }
+
+    pub(crate) fn is_sequencer_local(&self) -> bool {
+        matches!(self.sequencer, SequencerAccess::Local { .. })
+    }
+
+    pub(crate) fn known_global_tail(&self) -> TailOffsetWatch {
+        self.known_global_tail.clone()
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/loglet.rs:130]: It doesn't look like we need a clone at all call-sites. Maybe just return a reference and the caller can decide to clone if they need to?

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 5]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
                 _ = &mut cancel => {
                     break;
                 }
-                Some(_append) = self.append_stream.next() => {
+                Some(append) = self.append_stream.next() => {
+                    self.handle_append(Arc::clone(&provider), append).await;
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:67]: Do we need to clone the provider on every append?

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 6]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+            .rev()
+            .find(|segment| segment.index() == header.segment_index)
+            .ok_or(SequencerStatus::BadLogletId)?;
+
+        let loglet = provider
+            .get_replicated_loglet(header.log_id, header.segment_index, &segment.config.params)
+            .map_err(|err| SequencerStatus::Error(err.to_string()))?;
+
+        Ok(loglet)
+    }
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:128]: Can we get the loglet from the provider before we start the task? This way we won't need to clone the provider.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 7]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+{
+    fn get_loglet(
+        &self,
+        provider: &ReplicatedLogletProvider<T>,
+        header: &CommonRequestHeader,
+    ) -> Result<Arc<ReplicatedLoglet<T>>, SequencerStatus> {
+        let logs = self.metadata.logs();
+        let chain = logs
+            .chain(&header.log_id)
+            .ok_or(SequencerStatus::BadLogletId)?;
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:114]: `BadLogletId` implies that the id is malformed, but what we really mean is that we cannot find the loglet (`UnknownLogletId`).

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 8]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+
+        // appends are only possible to tail segment.
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:116]: The comment seems irrelevant.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 9]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+        let segment = chain
+            .iter()
+            .rev()
+            .find(|segment| segment.index() == header.segment_index)
+            .ok_or(SequencerStatus::BadLogletId)?;
+
+        let loglet = provider
+            .get_replicated_loglet(header.log_id, header.segment_index, &segment.config.params)
+            .map_err(|err| SequencerStatus::Error(err.to_string()))?;
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:125]: What is the error what we expect to happen here?

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 10]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+    async fn append(
+        &self,
+        provider: &ReplicatedLogletProvider<T>,
+        append: Append,
+    ) -> Result<Appended, SequencerStatus> {
+        let loglet = self.get_loglet(provider, &append.header)?;
+
+        // We get the sequencer instead of calling loglet.enqueue_batch to make sure this
+        // is the sequencer node. Otherwise, the client gets an error and should retry against
+        // the sequencer node.
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:170]: Comment is not in sync with code

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 11]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+        if !loglet.is_sequencer_local() {
+            return Err(SequencerStatus::NotSequencer);
+        }
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:173]: This also can be a check that happens even before we start the task if we move getting the loglet from provider to happen earlier.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 12]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+
+        let global_tail = loglet.known_global_tail();
+        // This is okay to be done in scope, since it's the only way to exert some back pressure on the callers
+        // until the sequencer has enough capacity to accept these records.
+        // This call should return immediately once records are scheduled for commit.
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:178]: Comment is out of sync with what's happening. We are performing the append in a task so we are not putting any back-pressure on the network channel.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 13]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+
+        // we spawn immediately to avoid slowing the request pump loop
+        // in case responses will go over a slow channel
+        let _ = task_center().spawn_child(
+            TaskKind::Disposable,
+            "handle-append-request",
+            None,
+            task.run(incoming),
+        );
+    }
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:94]: How do we uphold the original order of appends from the same sender with this approach? This appears to violate the ordering guarantees if we run tasks for appends coming from the same producer concurrently.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 14]
```diff
@@ -58,10 +63,164 @@ impl RequestPump {
...
+
+impl<T> HandleAppendTask<T>
+where
+    T: TransportConnect,
+{
+    fn get_loglet(
+        &self,
+        provider: &ReplicatedLogletProvider<T>,
+        header: &CommonRequestHeader,
+    ) -> Result<Arc<ReplicatedLoglet<T>>, SequencerStatus> {
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:110]: Why do we need to search the the metadata on every append? Can we try to find the loglet in active_loglets first by log_id/segment_id (the loglet_id should match the returned loglet) and only go down this search path if it's not already active?

File: crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs [hunk 15]
```diff
@@ -481,7 +505,7 @@ mod test {
                 header: CommonResponseHeader {
                     known_global_tail: None,
                     sealed: Some(false),
-                    status: self.reply_status,
+                    status: self.reply_status.clone(),
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs:508]: Is this needed?

**Review by AhmedSoliman — CHANGES_REQUESTED** [review]
Left a few comments. The most significant issue is that this violates the ordering guarantees for remote producers.

File: crates/bifrost/src/providers/replicated_loglet/network.rs (hunk 13 above)
Comment (muhamadazmy) [reply crates/bifrost/src/providers/replicated_loglet/network.rs:94]: Hmm, very good point. Before I refactor I actually only spawned after the records has already been queued. Thanks for catching this

File: crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs (hunk 15 above)
Comment (muhamadazmy) [reply crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs:508]: Yes, since SequencerStatus is no longer `Copy` because now it has the variant `Unknown(String)`

Comment (muhamadazmy) [reply crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs:508]: But I am wondering if we should expand the Unknow(string) to an inclusive list of status variants

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 16]
```diff
@@ -58,10 +107,174 @@ impl RequestPump {
...
+        let loglet = match provider.get_active_loglet(header.log_id, header.segment_index) {
+            Some(loglet) => loglet,
+            None => {
+                let logs = self.metadata.logs();
+                let chain = logs
+                    .chain(&header.log_id)
+                    .ok_or(SequencerStatus::UnknownLoglet)?;
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:167]: This is a different error. The LogId is unknown, not the loglet.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 17]
```diff
@@ -58,10 +107,174 @@ impl RequestPump {
...
+        let _ = task_center().spawn_child(TaskKind::Disposable, "wait-appended", None, task.run());
+    }
+
+    async fn get_loglet<T: TransportConnect>(
+        &self,
+        provider: &ReplicatedLogletProvider<T>,
+        header: &CommonRequestHeader,
+    ) -> Result<Arc<ReplicatedLoglet<T>>, SequencerStatus> {
+        let loglet = match provider.get_active_loglet(header.log_id, header.segment_index) {
+            Some(loglet) => loglet,
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:162]: It would be nice if you validate that this loglet has the loglet_id that we expect from the append header. Just as a sanity check.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 18]
```diff
@@ -58,10 +107,174 @@ impl RequestPump {
...
+
+impl WaitAppendedTask {
+    async fn run(self) -> anyhow::Result<()> {
+        let mut cancelled = std::pin::pin!(cancellation_watcher());
+
+        let result = tokio::select! {
+            _ = &mut cancelled => {
+                // we could try to respond to the request still
+                // but it's probably too late.
+                anyhow::bail!(ShutdownError);
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:224]: Disposable tasks will already be aborted on cancel so you don't need to watch cancellation here.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 19]
```diff
@@ -58,10 +107,174 @@ impl RequestPump {
...
+                    known_global_tail: Some(self.global_tail.latest_offset()),
+                    sealed: Some(self.global_tail.is_sealed()), // this must be true
+                    status: SequencerStatus::Sealed,
+                },
+                first_offset: LogletOffset::INVALID,
+            },
+            Err(AppendError::Shutdown(_)) => Appended {
+                header: CommonResponseHeader {
+                    known_global_tail: Some(self.global_tail.latest_offset()),
+                    sealed: Some(self.global_tail.is_sealed()), // this must be true
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:251]: comment is still wrong.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 20]
```diff
@@ -58,10 +107,174 @@ impl RequestPump {
...
+        };
+
+        tokio::select! {
+            sent = self.reciprocal.prepare(appended).send() => {
+                sent.map_err(|err| err.source)?;
+            },
+            _ = cancelled => {
+                anyhow::bail!(ShutdownError);
+            }
+        };
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:276]: No need for tokio select as stated above.

File: crates/bifrost/src/providers/replicated_loglet/network.rs [hunk 21]
```diff
@@ -58,10 +107,174 @@ impl RequestPump {
...
+                let logs = self.metadata.logs();
+                let chain = logs
+                    .chain(&header.log_id)
+                    .ok_or(SequencerStatus::UnknownLoglet)?;
+
+                let segment = chain
+                    .iter()
+                    .rev()
+                    .find(|segment| segment.index() == header.segment_index)
+                    .ok_or(SequencerStatus::UnknownLoglet)?;
```
Comment (AhmedSoliman) [inline crates/bifrost/src/providers/replicated_loglet/network.rs:173]: A nice future improvement is to pass the logs metadata version with the header and wait (or sync) the metadata to this version if (a) we didn't find the loglet and (b) our metadata is older.

**Review by AhmedSoliman — CHANGES_REQUESTED** [review]
We are very close from getting this to be ready, just a few tiny comments left.

**Comment by muhamadazmy** [comment]
@AhmedSoliman I applied your comments (again :D) please take another look

**Review by AhmedSoliman — APPROVED** [review]
